### PR TITLE
fs: fatfs: extend fatfs module cfg to support GPT and multi partitions

### PR DESCRIPTION
**Pull request by GrygiriiS**
This PR add two additional Kconfig options for **fatfs** module:
- CONFIG_FS_FATFS_LBA64 : enables support for 64-bit LBA, which also allows to enable  GUID Partition Table (GPT) support
- CONFIG_FS_MULTI_PARTITION : enables support for multiple volumes on the physical drive

This is required to support following use-case:
- platform is Raspberry Pi 5
- SD-card is formatted with GUID Partition Table (GPT) partition table
- FAT32 partition 0 is `bootfs` and should not be used by Zephyr app
- FAT32 partition 1 is allocated for Zephyr app

The sample  Zephyr app can be found in [1].

[1] https://github.com/xen-troops/zephyr-dom0-xt/tree/rpi5_dom0_dev

### CODE REVIEW

File: subsys/fs/Kconfig.fatfs [hunk 1]
```diff
@@ -230,6 +230,27 @@ config FS_FATFS_REENTRANT
...
 	  FatFs volume and a FatFs system mutex.
 
+config FS_FATFS_LBA64
+	bool "Support for 64-bit LBA"
+	depends on FS_FATFS_EXFAT
+	help
+	  This option enables support for 64-bit LBA, which also
+	  enables GPT support.
+
+config FS_MULTI_PARTITION
```
Comment (Laczen) [inline subsys/fs/Kconfig.fatfs:240]: Would it not be better to call this `FS_FATFS_GPT_PARTITIONS` ?

File: subsys/fs/Kconfig.fatfs [hunk 2]
```diff
@@ -230,6 +230,27 @@ config FS_FATFS_REENTRANT
...
+	bool "Support for multiple volumes on the physical drive"
+	help
```
Comment (Laczen) [inline subsys/fs/Kconfig.fatfs:242]: `depends on FS_FATFS_LBA64`

File: subsys/fs/Kconfig.fatfs [hunk 3]
```diff
@@ -230,6 +230,27 @@ config FS_FATFS_REENTRANT
...
+	  This option switches support for multiple volumes on the physical drive.
+	  By default (0), each logical drive number is bound to the same physical
```
Comment (Laczen) [inline subsys/fs/Kconfig.fatfs:244]: What about:
```	  
          When this function is enabled, each logical drive number can
	  be bound to arbitrary physical drive and partition listed
	  in the VolToPart[] of the fatfs module. The mount points
          have to be numbered in this case.
	  For example, 2 FAT partition on SD disk (3) in terms of Zephyr:
	   {3, 1} - mount point "/0:"
	   {3, 2} - mount point "/1:"
          When disabled (default), each logical drive number is bound to the same physical
	  drive number and only an FAT volume found on the physical drive will be
	  mounted. 
```

File: subsys/fs/Kconfig.fatfs (hunk 1 above)
Comment (GrygiriiS) [reply subsys/fs/Kconfig.fatfs:240]: It should work without GPT also, I'll change it to FS_FATFS_MULTI_PARTITION - ok?

Comment (Laczen) [reply subsys/fs/Kconfig.fatfs:240]: Yes, that would be ok.

Comment (GrygiriiS) [reply subsys/fs/Kconfig.fatfs]: fixed

File: subsys/fs/Kconfig.fatfs (hunk 3 above)
Comment (GrygiriiS) [reply subsys/fs/Kconfig.fatfs]: fixed

**Review by Laczen — APPROVED** [review]
LGTM, I'm wondering if we could add `VolToPart[]` as part of zephyr/zephyr fatfs module.

**Comment by GrygiriiS** [comment]
> LGTM, I'm wondering if we could add `VolToPart[]` as part of zephyr/zephyr fatfs module.

In my personal opinion It looks much less generic than, for example, `FF_VOLUME_STRS`, so I would recommend not doing this. Or, at least, not to do this until there are more real use cases implemented, which will allow to see if there is smth really generic which can be refactored.

**Comment by GrygiriiS** [comment]
Fix "Compliance Check" - removed `subsys` from commits subjects.

**Review by Laczen — DISMISSED** [review]

File: subsys/fs/Kconfig.fatfs [hunk 4]
```diff
@@ -230,6 +230,27 @@ config FS_FATFS_REENTRANT
...
+	help
+	  This option enables support for 64-bit LBA, which also
+	  enables GPT support.
+
+config FS_FATFS_MULTI_PARTITION
+	bool "Support for multiple volumes on the physical drive"
+	help
+	  When this function is enabled, each logical drive number can
+	  be bound to arbitrary physical drive and partition listed
+	  in the VolToPart[] of the fatfs module. The mount points
```
Comment (de-nordic) [inline subsys/fs/Kconfig.fatfs:245]: This is not clear from the description that this array has to be defined and provided by user. Just enabling the `FS_FATFS_MULTI_PARTITION` option will cause linker to fail to find the definition

Comment (GrygiriiS) [reply subsys/fs/Kconfig.fatfs:245]: Hi @de-nordic,

Would below change be ok with you:
```
--- a/subsys/fs/Kconfig.fatfs
+++ b/subsys/fs/Kconfig.fatfs
@@ -242,8 +242,9 @@ config FS_FATFS_MULTI_PARTITION
        help
          When this function is enabled, each logical drive number can
          be bound to arbitrary physical drive and partition listed
-         in the VolToPart[] of the fatfs module. The mount points
-         have to be numbered in this case.
+         in the VolToPart[] of the fatfs module. The VolToPart[] is expected to be
+         provided by Zephyr application.
+         The mount points have to be numbered in this case.
          For example, 2 FAT partition on SD disk (3) in terms of Zephyr:
            {3, 1} - mount point "/0:"
            {3, 2} - mount point "/1:"
```

Comment (de-nordic) [reply subsys/fs/Kconfig.fatfs:245]: Yeah. Would it be possible to provide test for this, serving also as a reference design?

Comment (GrygiriiS) [reply subsys/fs/Kconfig.fatfs:245]: Sorry, I'm not sure as it's required specific storage format and overall configuration is a bit specific.
How about adding code example in commit message:
```
Example of mounting a second FATFS partition places on SD-card:

static FATFS fat_fs;

static struct fs_mount_t mp = {
	.type = FS_FATFS,
	.fs_data = &fat_fs,
       .mnt_point = "/1:
};

/* 
 * 2 FAT partition on SD disk
 * PARTITION.pd - Physical drive number. "SD" has index 3 in terms of Zephyr (see FF_VOLUME_STRS)
 * PARTITION.pt - Partition (0:Auto detect, 1-4:Forced partition). So 1 for the first FATFS partition and 2 - for second.
 */
PARTITION VolToPart[FF_VOLUMES] = {
    [0] = {3, 1},     /* "0:" ==> 1st partition on the pd#0 */
    [1] = {3, 2},     /* "1:" ==> 2nd partition on the pd#0 */
    [2] = {0, -1},     /* "2:" ==> 3rd partition on the pd#0 */
    [3] = {0, -1},
    [4] = {0, -1},
    [5] = {0, -1},
    [6] = {0, -1},
    [7] = {0, -1},
};

fs_mount(&mp);
```

**Comment by GrygiriiS** [comment]
I'm very sorry for delayed response.
Changes:
- updated description of FS_FATFS_MULTI_PARTITION
- updated commit message and added code example.

File: subsys/fs/Kconfig.fatfs (hunk 4 above)
Comment (de-nordic) [reply subsys/fs/Kconfig.fatfs]: Seems that we would have to write MBR to flash partition, from within the test or sample.

**Review by de-nordic — APPROVED** [review]

**Comment by GrygiriiS** [comment]
Hi All,

Do I need to perform any actions here to have this PR merged?

**Comment by de-nordic** [comment]
> Hi All,
> 
> Do I need to perform any actions here to have this PR merged?

Second ACK. Post message, to Discard PR help channel, asking for reviews.

**Review by Laczen — APPROVED** [review]

**Pull request merged**